{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39d9de87",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import gseapy as gp\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53df72f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "org = gp.get_library_name(organism='Human')\n",
    "org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74363db5",
   "metadata": {},
   "outputs": [],
   "source": [
    "deg_file = '../deg/deg_results.xlsx'\n",
    "\n",
    "# Load the Excel file\n",
    "excel_file = pd.ExcelFile(deg_file)\n",
    "\n",
    "# Get the sheet names\n",
    "sheet_names = excel_file.sheet_names\n",
    "\n",
    "# Print the sheet names\n",
    "print(sheet_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d8f3ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# query genes\n",
    "# choose which pvalue column to use: pvalue or padj\n",
    "# set the threshold for the pvalue\n",
    "\n",
    "pval_col = 'padj'\n",
    "threshold = 0.05\n",
    "\n",
    "directory_path = f'{pval_col}_{threshold}'\n",
    "if not os.path.exists(directory_path):\n",
    "    # Directory does not exist, create it\n",
    "    os.makedirs(directory_path)\n",
    "    os.makedirs(f'{directory_path}/tables')\n",
    "    os.makedirs(f'{directory_path}/figures')\n",
    "    print(f\"Directory '{directory_path}' created successfully.\")\n",
    "\n",
    "symbol_col = 'external_gene_name'\n",
    "\n",
    "color_dict = {'GO_Molecular_Function_2021':'red', 'GO_Biological_Process_2021':'blue', \n",
    "                   'GO_Cellular_Component_2021':'green','KEGG_2021_Human':'yellow'}\n",
    "\n",
    "for pval_col,threshold in zip(['padj'],[0.05]):\n",
    "\n",
    "    for sheet_name in sheet_names:\n",
    "        print(sheet_name)\n",
    "        \n",
    "        df = pd.read_excel(deg_file, sheet_name=sheet_name)\n",
    "        df = df[df[symbol_col].notna()]\n",
    "        \n",
    "        # Load the list of differentially expressed genes\n",
    "        diff_exp_genes = df[df[pval_col]<threshold][symbol_col].tolist()\n",
    "        if len(diff_exp_genes)==0:\n",
    "            print('No DEGs.')\n",
    "            continue\n",
    "\n",
    "        # Load the background gene set\n",
    "        background_genes =  df[df[pval_col].notna()][symbol_col].tolist()\n",
    "\n",
    "        # Set up the Enrichr query parameters\n",
    "        enrichr_results = gp.enrichr(\n",
    "            gene_list=diff_exp_genes,\n",
    "            background=background_genes,\n",
    "            organism='Human',\n",
    "            gene_sets=['GO_Molecular_Function_2021', 'GO_Biological_Process_2021', \n",
    "                    'GO_Cellular_Component_2021','KEGG_2021_Human', 'Human_Gene_Atlas',\n",
    "                    'Descartes_Cell_Types_and_Tissue_2021','ARCHS4_Tissues','GTEx_Tissue_Expression_Down',\n",
    "                    'GTEx_Tissue_Expression_Up','GTEx_Tissues_V8_2023','Tissue_Protein_Expression_from_ProteomicsDB'\n",
    "                    ],\n",
    "            outdir='enrichr_results',\n",
    "            cutoff=1\n",
    "        )\n",
    "            \n",
    "        df = enrichr_results.results\n",
    "        #filter enrichment results for adjusted p-value < 0.05\n",
    "        df = df[df['Adjusted P-value']<0.05]\n",
    "        if len(df)==0:\n",
    "            print('No enrichment results.')\n",
    "            continue\n",
    "        df = df.sort_values(by='Adjusted P-value', ascending=False)\n",
    "        \n",
    "        #horizontal bar plot\n",
    "        fig, ax = plt.subplots(figsize=(20, 10))\n",
    "        gene_sets = sorted(list(set(df.Gene_set)))\n",
    "\n",
    "        # create an ExcelWriter object to save results\n",
    "        writer = pd.ExcelWriter(f'{pval_col}_{threshold}/tables/{sheet_name}.xlsx', engine='xlsxwriter')\n",
    "\n",
    "        for gs in gene_sets:\n",
    "            gs_df = df[df.Gene_set==gs]\n",
    "            gs_df = gs_df.sort_values(by=['Adjusted P-value', 'Odds Ratio'], ascending=[True, False])\n",
    "            gs_df = gs_df.reset_index(drop=True)\n",
    "            \n",
    "            if gs in ['GO_Molecular_Function_2021', 'GO_Biological_Process_2021', \n",
    "                    'GO_Cellular_Component_2021','KEGG_2021_Human']:\n",
    "                # invert order for plotting \n",
    "                plot_gs_df = gs_df.copy()\n",
    "                plot_gs_df = plot_gs_df.iloc[:5,:]\n",
    "                plot_gs_df = plot_gs_df.iloc[::-1]\n",
    "                ax.barh(y=plot_gs_df.Term, width=-np.log10(plot_gs_df['Adjusted P-value']), color=color_dict[gs], label=gs)\n",
    "\n",
    "            #sheetname has to be shorter than 31 characters\n",
    "            if len(gs) > 31:\n",
    "                gs = gs[:31]\n",
    "            gs_df.to_excel(writer, sheet_name=f'{gs}',index=False)\n",
    "            \n",
    "        ax.set_xlabel('-log10(adjusted p-value)', size=20)\n",
    "        ax.yaxis.set_tick_params(labelsize=15)\n",
    "        ax.xaxis.set_tick_params(labelsize=14)\n",
    "        ax.set_title(sheet_name, size=24)\n",
    "        ax.legend(loc='lower right', fontsize=15)\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(f'{pval_col}_{threshold}/figures/{sheet_name}.png')\n",
    "        \n",
    "        # save the Excel file\n",
    "        writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "074edc56",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Allen Brain Up/Down - query only up or down genes\n",
    "# choose which pvalue column to use: pvalue or padj\n",
    "# set the threshold for the pvalue\n",
    "\n",
    "pval_col = 'padj'\n",
    "threshold = 0.05\n",
    "\n",
    "directory_path = f'{pval_col}_{threshold}'\n",
    "if not os.path.exists(f'{directory_path}/tables/brain/'):\n",
    "    os.makedirs(f'{directory_path}/tables/brain/')\n",
    "if not os.path.exists(f'{directory_path}/figures/brain/'):\n",
    "    os.makedirs(f'{directory_path}/figures/brain/')\n",
    "\n",
    "symbol_col = 'external_gene_name'\n",
    "\n",
    "color_dict = {'Allen_Brain_Atlas_up':'red', 'Allen_Brain_Atlas_down':'blue'}\n",
    "\n",
    "for pval_col,threshold in zip(['padj'],[0.05]):\n",
    "\n",
    "\n",
    "    for sheet_name in sheet_names:\n",
    "        print(sheet_name)\n",
    "        \n",
    "        df = pd.read_excel(deg_file, sheet_name=sheet_name)\n",
    "        df = df[df[symbol_col].notna()]\n",
    "        \n",
    "        # Load the list of differentially expressed genes\n",
    "        up_de_genes = df[(df[pval_col]<threshold) & (df.log2FoldChange>0)][symbol_col].tolist()\n",
    "        down_de_genes = df[(df[pval_col]<threshold) & (df.log2FoldChange<0)][symbol_col].tolist()\n",
    "        \n",
    "        if len(up_de_genes)==0 and len(down_de_genes)==0:\n",
    "            print('No DEGs.')\n",
    "            continue\n",
    "\n",
    "        # Load the background gene set\n",
    "        background_genes =  df[df[pval_col].notna()][symbol_col].tolist()\n",
    "\n",
    "        # Set up the Enrichr query parameters\n",
    "        up_enrichr_results = gp.enrichr(\n",
    "            gene_list=up_de_genes,\n",
    "            background=background_genes,\n",
    "            organism='Human',\n",
    "            gene_sets=['Allen_Brain_Atlas_up'],\n",
    "            outdir='enrichr_results',\n",
    "            cutoff=1\n",
    "        )\n",
    "        \n",
    "        down_enrichr_results = gp.enrichr(\n",
    "            gene_list=down_de_genes,\n",
    "            background=background_genes,\n",
    "            organism='Human',\n",
    "            gene_sets=['Allen_Brain_Atlas_down'],\n",
    "            outdir='enrichr_results',\n",
    "            cutoff=1\n",
    "        )\n",
    "            \n",
    "        up_df = up_enrichr_results.results\n",
    "        up_df = up_df[up_df['Adjusted P-value']<0.05]\n",
    "        \n",
    "        down_df = down_enrichr_results.results\n",
    "        down_df = down_df[down_df['Adjusted P-value']<0.05]\n",
    "        \n",
    "        if up_df.shape[0]==0 and down_df.shape[0]==0:\n",
    "            print('No enrichment results.')\n",
    "            continue\n",
    "        \n",
    "        df = pd.concat([up_df, down_df])\n",
    "        df.loc[df.Gene_set=='Allen_Brain_Atlas_up', 'color'] = 'red'\n",
    "        df.loc[df.Gene_set=='Allen_Brain_Atlas_down', 'color'] = 'blue'\n",
    "        df = df.sort_values(by='Adjusted P-value', ascending=False)\n",
    "\n",
    "        #horizontal bar plot\n",
    "        fig, ax = plt.subplots(figsize=(20, 10))\n",
    "        gene_sets = sorted(list(set(df.Gene_set)))\n",
    "        for gs in gene_sets:\n",
    "            gs_df = df[df.Gene_set==gs]\n",
    "            gs_df = gs_df.iloc[:5,:]\n",
    "            ax.barh(y=gs_df.Term, width=-np.log10(gs_df['Adjusted P-value']), color=color_dict[gs], label=gs)\n",
    "            \n",
    "        ax.set_xlabel('-log10(adjusted p-value)', size=20)\n",
    "        ax.yaxis.set_tick_params(labelsize=15)\n",
    "        ax.xaxis.set_tick_params(labelsize=14)\n",
    "        ax.set_title(sheet_name, size=24)\n",
    "        ax.legend(loc='lower right', fontsize=15)\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(f'{pval_col}_{threshold}/figures/brain/{sheet_name}.png')\n",
    "        \n",
    "        # create an ExcelWriter object\n",
    "        writer = pd.ExcelWriter(f'{pval_col}_{threshold}/tables/brain/{sheet_name}.xlsx', engine='xlsxwriter')\n",
    "        \n",
    "        for gs in set(df.Gene_set):\n",
    "            # write each data frame to a separate sheet\n",
    "            gs_df = df[df.Gene_set==gs]\n",
    "            if len(gs) > 31:\n",
    "                gs = gs[:31]\n",
    "            gs_df.to_excel(writer, sheet_name=f'{gs}',index=False)\n",
    "\n",
    "        # save the Excel file\n",
    "        writer.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
